plugins/gimp/file-jxl-load.cc

```diff
@@ -10,163 +10,252 @@
+        } else {
+          image_type = GIMP_RGB;
+          layer_type = GIMP_RGBA_IMAGE;
+          format.num_channels = info.num_color_channels + 1;
+        }
+      } else if (info.num_color_channels == 1) {
+        if (info.alpha_bits == 0) {
+          image_type = GIMP_GRAY;
+          layer_type = GIMP_GRAY_IMAGE;
+          format.num_channels = info.num_color_channels;
+        } else {
+          image_type = GIMP_GRAY;
+          layer_type = GIMP_GRAYA_IMAGE;
+          format.num_channels = info.num_color_channels + 1;
+        }
+      }
+
+      if (info.bits_per_sample <= 8) {
+        if (loadlinear) {
+          format.data_type = JXL_TYPE_UINT8;
+          precision = GIMP_PRECISION_U8_LINEAR;
+        } else {
+          format.data_type = JXL_TYPE_UINT8;
+          precision = GIMP_PRECISION_U8_GAMMA;
+        }
+      } else if (info.bits_per_sample <= 16) {
+        if (info.exponent_bits_per_sample > 0) {
+          if (loadlinear) {
+            format.data_type = JXL_TYPE_FLOAT16;
+            precision = GIMP_PRECISION_HALF_LINEAR;
+          } else {
+            format.data_type = JXL_TYPE_FLOAT16;
+            precision = GIMP_PRECISION_HALF_GAMMA;
+          }
+        } else if (loadlinear) {
+          format.data_type = JXL_TYPE_UINT16;
+          precision = GIMP_PRECISION_U16_LINEAR;
+        } else {
+          format.data_type = JXL_TYPE_UINT16;
+          precision = GIMP_PRECISION_U16_GAMMA;
+        }
+      } else {
+        if (info.exponent_bits_per_sample > 0) {
+          if (loadlinear) {
+            format.data_type = JXL_TYPE_FLOAT;
+            precision = GIMP_PRECISION_FLOAT_LINEAR;
+          } else {
+            format.data_type = JXL_TYPE_FLOAT;
+            precision = GIMP_PRECISION_FLOAT_GAMMA;
+          }
+        } else if (loadlinear) {
+          format.data_type = JXL_TYPE_UINT32;
+          precision = GIMP_PRECISION_U32_LINEAR;
+        } else {
+          format.data_type = JXL_TYPE_UINT32;
+          precision = GIMP_PRECISION_U32_GAMMA;
+        }
+      }
+    } else if (status == JXL_DEC_COLOR_ENCODING) {
+      // Get the ICC color profile of the pixel data
+      size_t icc_size;
+      if (JXL_DEC_SUCCESS !=
+          JxlDecoderGetICCProfileSize(
+              dec.get(), &format, JXL_COLOR_PROFILE_TARGET_DATA, &icc_size)) {
+        g_printerr("Error: JxlDecoderGetICCProfileSize failed\n");
+        return false;
+      }
+
+      if (icc_size > 0) {
+        icc_profile.resize(icc_size);
+        if (JXL_DEC_SUCCESS != JxlDecoderGetColorAsICCProfile(
+                                   dec.get(), &format,
+                                   JXL_COLOR_PROFILE_TARGET_DATA,
+                                   icc_profile.data(), icc_profile.size())) {
+          g_printerr("Error: JxlDecoderGetColorAsICCProfile failed\n");
+          return false;
+        }
+
+        GError **error = NULL;
+        profile = gimp_color_profile_new_from_icc_profile(
```

Comment:
I think this will reintroduce https://gitlab.com/wg1/jpeg-xl/-/issues/182. Ideally, we would also try to get the color profile as a `JxlColorEncoding` and, if successful, compare it against sRGB and use `gimp_color_profile_new_rgb_srgb` if it matches. Would you like to give it a try in this pull request or should we do it later?

Reply:
I didn't notice the regression because I had GIMP set to keep embedded profiles.  The problem is Thread Pool Internal has access to CodecInOut, but Parallel Runner doesn't, as far as I can tell.  I'll take a look at it.

Reply:
I think the regression is fixed now...  it's a bit complicated because I'm avoiding using internal headers.  (I'd like to remove "lib/jxl/base/file_io.h" someday.)